stream/src/connection/mod.rs

```diff
@@ -0,0 +1,147 @@
+//!
+//! To better protect against malicious peers that create and/or accept connections but do not participate in handshakes,
+//! a configurable deadline is enforced for any handshake to be completed. This allows for the underlying runtime to maintain
+//! a standard read/write timeout for connections without making it easier for malicious peers to keep useless connections open.
+//!
+//! ## Encryption
+//!
+//! During the handshake (described above), a shared x25519 secret is established using a Diffie-Hellman Key Exchange. This
+//! x25519 secret is then used to create a ChaCha20-Poly1305 cipher for encrypting all messages exchanged with the peer.
+//!
+//! ChaCha20-Poly1305 nonces (12 bytes) are constructed such that the first bit indicates whether the sender is a dialer (1) or
+//! dialee (0). The rest of the first byte (next 7 bits) and next byte (all 8 bits) are unused (set to 0). The next 2 bytes
+//! are a `u16` iterator and the final 8 bytes are a `u64` sequence number. When the sequence reaches `u64::MAX`, the iterator
+//! is incremented and the sequence is reset to 0. This technique provides each sender with a channel duration of `2^80` frames
+//! (and automatically terminates when this number of frames has been sent). In the blockchain context, validators often maintain
+//! long-lived connections with each other and avoiding connection re-establishment (to reset iterator/sequence with a new cipher)
+//! is desirable.
+//!
+//! ```text
+//! +---+---+---+---+---+---+---+---+---+---+---+---+
+//! | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |10 |11 |
+//! +---+---+---+---+---+---+---+---+---+---+---+---+
+//! | D | U |It(u16)|         Sequence(u64)         |
+//! +---+---+---+---+---+---+---+---+---+---+---+---+
+//!
+//! D = Dialer/Dialee, U = Unused, It = Iterator
+//! ```
+//!
+//! _We use a combination of `u64` (sequence) and `u16` (iterator) instead of implementing `u80/u88` because
+//! CPUs provide native support for `u64` operations (which will always be faster than an implementation of a
+//! "wrapping add" over arbitrary bytes). With this technique, almost all operations (other than iterator
+//! increments every `2^64` frames) are just a basic `u64` increment._
+//!
+//! This simple coordination prevents nonce reuse (which would allow for messages to be decrypted) and saves a small amount of
+//! bandwidth (no need to send the nonce alongside the encrypted message). This "pedantic" construction of the nonce
+//! also avoids accidentally reusing a nonce over long-lived connections when setting it to be a small hash (as in XChaCha-Poly1305).
+
+use commonware_cryptography::Scheme;
+use prost::DecodeError;
+use std::time::Duration;
+use thiserror::Error;
+
+mod handshake;
+mod instance;
+mod utils;
+mod wire {
+    include!(concat!(env!("OUT_DIR"), "/wire.rs"));
+}
+mod x25519;
+
+pub use handshake::IncomingHandshake;
+pub use instance::{Instance, Sender};
+
+#[derive(Clone)]
+pub struct Config<C: Scheme> {
```

Comment:
Will probably also want to accept metrics here?

Reply:
Fixed by https://github.com/commonwarexyz/monorepo/issues/184